packages/flutter/lib/src/gestures/tap_and_drag.dart

```diff
@@ -1139,25 +1160,35 @@ sealed class BaseTapAndDragGestureRecognizer extends OneSequenceGestureRecognize
   String get debugDescription => 'tap_and_drag';
 
   void _acceptDrag(PointerEvent event) {
+    assert(_dragState == _DragState.accepted);
+
     if (!_wonArenaForPrimaryPointer) {
       return;
     }
+
     if (dragStartBehavior == DragStartBehavior.start) {
-      _initialPosition = _initialPosition + OffsetPair(global: event.delta, local: event.localDelta);
+      _initialPosition += OffsetPair(
+        global: event.delta,
+        local: event.localDelta,
+      );
+      _currentPosition = _initialPosition;
     }
     _checkDragStart(event);
-    if (event.localDelta != Offset.zero) {
-      final Matrix4? localToGlobal = event.transform != null ? Matrix4.tryInvert(event.transform!) : null;
-      final Offset correctedLocalPosition = _initialPosition.local + event.localDelta;
+    final Offset localDelta = event.localDelta;
+    if (localDelta != Offset.zero) {
+      _currentPosition = OffsetPair.fromEventPosition(event);
+      final Offset correctedLocalPosition = _initialPosition.local + localDelta;
+      final Matrix4? localToGlobalTransform = event.transform == null ? null : Matrix4.tryInvert(event.transform!);
       final Offset globalUpdateDelta = PointerEvent.transformDeltaViaPositions(
+        transform: localToGlobalTransform,
+        untransformedDelta: localDelta,
         untransformedEndPosition: correctedLocalPosition,
-        untransformedDelta: event.localDelta,
-        transform: localToGlobal,
       );
-      final OffsetPair updateDelta = OffsetPair(local: event.localDelta, global: globalUpdateDelta);
-      _correctedPosition = _initialPosition + updateDelta; // Only adds delta for down behaviour
-      _checkDragUpdate(event);
-      _correctedPosition = null;
+      final OffsetPair updateDelta = OffsetPair(
+        local: localDelta,
+        global: globalUpdateDelta,
+      );
+      _checkDragUpdate(event, corrected: _initialPosition + updateDelta);
```

Comment:
nit: I think we should add the comment back regarding only adding the delta for down behavior.